avl-tree/src/lib.rs

```diff
@@ -0,0 +1,589 @@
+            // path to the leftmost descendant
+            let mut inner_path = Vec::with_capacity((self.size() as f64).log2() as usize);
+
+            while self.get_field(leftmost, Field::Left) != SENTINEL {
+                leftmost_parent = leftmost;
+                leftmost = self.get_field(leftmost, Field::Left);
+                inner_path.push((Some(leftmost_parent), Some(Field::Left), leftmost));
+            }
+            
+            if leftmost_parent != SENTINEL {
+                self.set_field(
+                    leftmost_parent,
+                    Field::Left,
+                    self.get_field(leftmost, Field::Right),
+                );
+            }
+
+            self.set_field(leftmost, Field::Left, left);
+            if right != leftmost {
+                self.set_field(leftmost, Field::Right, right);
+            }
+
+            let (parent, branch, _) = path.pop().unwrap();
+
+            if let Some(parent) = parent {
+                self.set_field(parent, branch.unwrap(), leftmost);
+            }
+
+            path.push((parent, branch, leftmost));
+            if right != leftmost {
+                path.push((Some(leftmost), Some(Field::Right), right));
+            }
+            // drop the last inner_path element since it references the leftmost node
+            if !inner_path.is_empty() {
+                inner_path.pop();
+            }
+            path.extend(inner_path);
+
+            leftmost
+        } else {
+            let child = if left == SENTINEL && right == SENTINEL {
+                SENTINEL
+            } else {
+                if left != SENTINEL { left } else { right }
+            };
+
+            let (parent, branch, _) = path.pop().unwrap();
+
+            if let Some(parent) = parent {
+                self.set_field(parent, branch.unwrap(), child);
+
+                if child != SENTINEL {
+                    path.push((Some(parent), branch, child));
+                }
+            }
+
+            child
+        };
+
+        if node_index == self.root {
+            self.root = replacement;
+        }
+
+        self.delete(node_index);
+        self.rebalance(path);
+
+        Some(value)
+    }
+
+    fn balance_factor(&self, left: u32, right: u32) -> i32 {
+        // safe to convert to i32 since height will be at most log2(capacity)
+        let left_height = if left != SENTINEL {
+            self.get_field(left, Field::Height) as i32 + 1
+        } else {
+            0
+        };
+        let right_height = if right != SENTINEL {
+            self.get_field(right, Field::Height) as i32 + 1
+        } else {
+            0
+        };
+
+        left_height - right_height
+    }
+
+    fn left_rotate(&mut self, index: u32) -> u32 {
+        let right = self.get_field(index, Field::Right);
+        let right_left = self.get_field(right, Field::Left);
+
+        self.set_field(index, Field::Right, right_left);
+        self.set_field(right, Field::Left, index);
```

Comment:
This is a question more for my understanding: why are the respective heights of each node not updated in this function? At first glance it looks like the pointers get misaligned, but I'm also not sure what you do after calling these helpers:


```
      I
         \
           R
          /
       RL
```
becomes

```
          R
        /
      I
         \
           RL
```

But what happens to the previous parent of `index`? There's still going to be a dangling pointer, unless this is handled elsewhere. Additionally, it seems like the heights of `index` and `right` should be swapped

Reply:
Additionally, I don't understand how this decreases the height of the tree? Not well versed on AVL, but my general understanding of rotations is that they should reduce the overall tree height in certain situations.

Reply:
The `*-rotate` functions do not update the parent of the `index` node as the parent is not passed as a parameter. They return the reference to the top-most node after the rotation, which then is used to update the parent. In most cases, a rotation decreases the height of the tree, but some rotations are used in preparation to another one and might not decrease the height at first – e.g., left-right rotation:
```
    Z                                Z                                   Y
   / \                              / \                                /   \
  X   d   (left rotation on X) =>  Y   d   (right rotation on Z) =>  X       Z
 / \                              / \                               / \     / \
a   Y                            X   c                             a   b   c   d
   / \                          / \
  b   c                        a   b
```
(`a` `b`, `c` and `d` can be other sub-trees, not necessarily leaf nodes)

The height was updated when the node's balance factor was between `-1` and `1` in the `rebalance` function, which means that there could be cases where the height stored in a register does not precisely reflect the height of the (sub-)tree. I changed the `set_field` function to update the height every time a new child is set. Perhaps there could be edge cases that would require the height of the node to up to date on the `height` register – although I could not come up with one. It does not seem to impact the performance.